crates/keccak/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+use bellpepper_core::boolean::Boolean;
+use bellpepper_core::ConstraintSystem;
+use bellpepper_core::SynthesisError;
+
+mod uint64;
+use ff::PrimeField;
+use uint64::UInt64;
+
+#[rustfmt::skip]
+const ROUND_CONSTANTS: [u64; 24] = [
+    0x0000000000000001, 0x0000000000008082, 0x800000000000808a, 0x8000000080008000,
+    0x000000000000808b, 0x0000000080000001, 0x8000000080008081, 0x8000000000008009,
+    0x000000000000008a, 0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
+    0x000000008000808b, 0x800000000000008b, 0x8000000000008089, 0x8000000000008003,
+    0x8000000000008002, 0x8000000000000080, 0x000000000000800a, 0x800000008000000a,
+    0x8000000080008081, 0x8000000000008080, 0x0000000080000001, 0x8000000080008008,
+];
+const ROTR: [usize; 25] = [
+    0, 1, 62, 28, 27, 36, 44, 6, 55, 20, 3, 10, 43, 25, 39, 41, 45, 15, 21, 8, 18, 2, 61, 56, 14,
+];
+
+fn xor_2<E, CS>(mut cs: CS, a: &UInt64, b: &UInt64) -> Result<UInt64, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    // a ^ b
+    a.xor(cs.namespace(|| "xor_2"), b)
+}
+
+fn xor_5<E, CS>(
+    mut cs: CS,
+    a: &UInt64,
+    b: &UInt64,
+    c: &UInt64,
+    d: &UInt64,
+    e: &UInt64,
+) -> Result<UInt64, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    // a ^ b ^ c ^ d ^ e
+    let ab = a.xor(cs.namespace(|| "xor_5 first"), b)?;
+    let abc = ab.xor(cs.namespace(|| "xor_5 second"), c)?;
+    let abcd = abc.xor(cs.namespace(|| "xor_5 third"), d)?;
+    abcd.xor(cs.namespace(|| "xor_5 fourth"), e)
+}
+
+fn xor_not_and<E, CS>(
+    mut cs: CS,
+    a: &UInt64,
+    b: &UInt64,
+    c: &UInt64,
+) -> Result<UInt64, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    // a ^ ((!b) & c)
+    let nb = b.not();
+    let nbc = nb.and(cs.namespace(|| "xor_not_and second"), c)?;
+    a.xor(cs.namespace(|| "xor_not_and third"), &nbc)
+}
+
+fn round_1600<E, CS>(mut cs: CS, a: Vec<UInt64>, rc: u64) -> Result<Vec<UInt64>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+{
+    assert_eq!(a.len(), 25);
+
+    // # θ step
+    // C[x] = A[x,0] xor A[x,1] xor A[x,2] xor A[x,3] xor A[x,4],   for x in 0…4
+    let mut c = Vec::new();
+    for x in 0..5 {
+        let cs = &mut cs.namespace(|| format!("omega c {}", x));
+
+        c.push(xor_5(
+            cs,
+            &a[x],
+            &a[x + 5usize],
+            &a[x + 10usize],
+            &a[x + 15usize],
+            &a[x + 20usize],
```

Comment:
These `usize` type specifiers here and throughout this file seem unnecessary and just make reading/interpreting the algorithm harder.

Reply:
I think they're voluntary: the goal is to make it clear the operand is on an index variable, not on something supposed to represent one of the u64 or Uint64 being modeled here.

Reply:
It has the opposite effect for me, since without the annotation this should be fine iff `x` is of the right type. The annotation makes it feel like some extra work is being done to coerce (although I know that's not the exact semantics either).